compiler/rustc_hir_typeck/src/cast.rs

```diff
@@ -501,12 +501,22 @@ impl<'a, 'tcx> CastCheck<'tcx> {
                             .must_apply_modulo_regions()
                         {
                             label = false;
-                            err.span_suggestion(
-                                self.span,
-                                "consider using the `From` trait instead",
-                                format!("{}::from({})", self.cast_ty, snippet),
-                                Applicability::MaybeIncorrect,
-                            );
+                            if !self.cast_ty.is_primitive() {
+                                let type_to_cast = self.cast_ty.to_string().replacen('<', "::<", 1);
```

Comment:
```suggestion
                                let ty_path_str = if let ty::Adt(def, args) = self.cast_ty.kind() {
                                    fcx.tcx.value_path_str_with_args(def.did(), args)
                                } else {
                                    self.cast_ty.to_string()
                                };
};
```
You can also use `Diag::span_suggestion_with_style` to dedup the code below, i.e.
```rs
let style = if self.cast_ty.is_primitive() {
    SuggestionStyle::ShowCode
} else {
    SuggestionStyle::ShowAlways
};
err.span_suggestion_with_style(
    ...,
    style,
);
```

Reply:
This is a cold path (i.e. only reachable if the user code is erroneous), so it's more beneficial to choose the style you prefer rather than the faster one. Stringifying types or calling `str::replacen` is okay, my suggestion is more because there may be future types where replacing `<` with `::<` is wrong, and to avoid reinventing the wheel.

Performance-wise, `Ty::to_string` and `TyCtx::value_path_str_with_args` ends up calling the same function for ADTs, while `str::replacen` requires reallocating a string.
https://github.com/rust-lang/rust/blob/553600e0f5f5a7d492de6d95ccb2f057005f5651/compiler/rustc_middle/src/ty/print/pretty.rs#L803-L805
https://github.com/rust-lang/rust/blob/553600e0f5f5a7d492de6d95ccb2f057005f5651/compiler/rustc_middle/src/ty/print/pretty.rs#L2458-L2460

Reply:
Hmm, good point, thanks for showing me this method I will use it